ipc/provider/src/manager/evm/gas_estimator_middleware.rs

```diff
@@ -0,0 +1,160 @@
+            Eip1559GasEstimatorError::MiddlewareError(e) => Some(e),
+            _ => None,
+        }
+    }
+}
+
+pub struct GasFeeEstimate {
+    pub max_priority_fee_per_gas: U256,
+    pub max_fee_per_gas: U256,
+}
+
+#[derive(Debug)]
+pub struct Eip1559GasEstimatorMiddleware<M: Middleware> {
+    inner: M,
+}
+
+impl<M: Middleware> Eip1559GasEstimatorMiddleware<M> {
+    pub fn new(inner: M) -> Self {
+        Self { inner }
+    }
+
+    pub async fn max_priority_fee_per_gas(&self) -> Result<U256, ProviderError> {
+        self.inner
+            .provider()
+            .request("eth_maxPriorityFeePerGas", json!([]))
+            .await
+    }
+
+    pub async fn base_fee_per_gas(&self) -> Result<U256, ProviderError> {
+        let latest_block = self
+            .inner
+            .provider()
+            .get_block(ethers::types::BlockNumber::Latest)
+            .await?
+            .ok_or_else(|| ProviderError::CustomError("Latest block not found".to_string()))?;
+
+        latest_block
+            .base_fee_per_gas
+            .ok_or_else(|| ProviderError::CustomError("EIP-1559 not activated".to_string()))
+    }
+
+    pub async fn estimate_gas_fees(&self) -> Result<GasFeeEstimate, Eip1559GasEstimatorError<M>> {
+        let max_priority_fee_per_gas = self
+            .max_priority_fee_per_gas()
+            .await
+            .map_err(|e| Eip1559GasEstimatorError::FailedToEstimateGas(e.to_string()))?;
+        let base_fee_per_gas = self
+            .base_fee_per_gas()
+            .await
+            .map_err(|e| Eip1559GasEstimatorError::FailedToEstimateGas(e.to_string()))?;
+
+        // Buffer the base fee by multiplying by 2 to account for potential cumulative increases.
+        let base_fee_per_gas_surged = base_fee_per_gas * 2;
+        let max_fee_per_gas = max_priority_fee_per_gas + base_fee_per_gas_surged;
+
+        Ok(GasFeeEstimate {
+            max_priority_fee_per_gas,
+            max_fee_per_gas,
+        })
+    }
+}
+
+#[async_trait]
+impl<M: Middleware> Middleware for Eip1559GasEstimatorMiddleware<M> {
+    type Error = Eip1559GasEstimatorError<M>;
+    type Provider = M::Provider;
+    type Inner = M;
+
+    fn inner(&self) -> &M {
+        &self.inner
+    }
+
+    /// Fills the transaction with EIP-1559 gas fees.
+    async fn fill_transaction(
+        &self,
+        tx: &mut TypedTransaction,
+        block: Option<BlockId>,
+    ) -> Result<(), Self::Error> {
+        if let TypedTransaction::Eip1559(inner) = tx {
+            let gas_fees = self.estimate_gas_fees().await?;
+
+            // Set the gas fees directly on the transaction.
+            let tx_req = inner
+                .clone()
+                .max_fee_per_gas(gas_fees.max_fee_per_gas)
+                .max_priority_fee_per_gas(gas_fees.max_priority_fee_per_gas);
+
+            *tx = TypedTransaction::Eip1559(tx_req);
+        } else {
+            return Err(Eip1559GasEstimatorError::failed_to_estimate_gas_not_supported());
```

Comment:
why not just skip the error so that other types of txns can go through so that it's more general. Or the purpose is only supporting EIP1559?

Reply:
> Or the purpose is only supporting EIP1559?

Yes. As these endpoints are only available from EIP1559 upgrade.